spot_rl_experiments/spot_rl/envs/place_env.py

```diff
@@ -2,69 +2,216 @@
+                dir_rpy_to_goal = get_RPY_from_vector(goal_ee_pos - curr_ee_pos)
+
+                # Go to intermediate point
+                cmd_id = self.spot.move_gripper_to_point(
+                    intr_ee_pos,
+                    dir_rpy_to_goal,
+                    self.config.ARM_TRAJECTORY_TIME_IN_SECONDS,
+                )
+                self.spot.block_until_arm_arrives(cmd_id, timeout_sec=10)
+
+                # Direct the gripper to face downwards
+                dir_rpy_to_goal = [0.0, np.pi / 2, 0.0]
+
+                # Go to goal point
+                cmd_id = self.spot.move_gripper_to_point(
+                    goal_ee_pos,
+                    dir_rpy_to_goal,
+                    self.config.ARM_TRAJECTORY_TIME_IN_SECONDS,
+                )
+                self.spot.block_until_arm_arrives(cmd_id, timeout_sec=10)
+
+            success_list.append(
+                {
+                    "time_taken": time.time() - start_time,
+                    "success": self.place_env.place_attempted or not self.use_policies,
+                    "place_target": self.place_env.get_corrected_place_target_in_base_frame(),
+                    "ee_pos": self.place_env.get_gripper_position_in_base_frame(),
+                }
+            )
+
+            # Open gripper to drop the object
+            self.spot.open_gripper()
+            # Add sleep as open_gripper() is a non-blocking call
+            time.sleep(1)
+
+        return success_list
+
+    def shutdown(self, should_dock=False):
+        try:
+            if should_dock:
+                self.place_env.say("Executing automatic docking")
+                dock_start_time = time.time()
+                while time.time() - dock_start_time < 2:
+                    try:
+                        self.spot.dock(dock_id=DOCK_ID, home_robot=True)
+                    except Exception:
+                        print("Dock not found... trying again")
+                        time.sleep(0.1)
+            else:
+                self.place_env.say("Will sit down here")
+                self.spot.sit()
+        finally:
+            self.spot.power_off()
 
 
 class SpotPlaceEnv(SpotBaseEnv):
-    def __init__(self, config, spot: Spot, place_target, target_is_local=False):
+    def __init__(self, config, spot: Spot):
         super().__init__(config, spot)
-        self.place_target = np.array(place_target)
-        self.place_target_is_local = target_is_local
+        self.place_target = None
+        self.place_target_is_local = False
+
         self.ee_gripper_offset = mn.Vector3(config.EE_GRIPPER_OFFSET)
         self.placed = False
 
-    def reset(self, *args, **kwargs):
+    def reset(self, place_target, target_is_local=False, *args, **kwargs):
+        assert place_target is not None
+        self.place_target = np.array(place_target)
+        self.place_target_is_local = target_is_local
+
         # Move arm to initial configuration
+        self.spot.close_gripper()
```

Comment:
um why are we closing the gripper? Wondering if this works well with the ASC pipeline

Reply:
ASC does not use `SpotPlaceEnv` at all.
ASC creates its own environment